lib/smack/SmackRep.cpp

```diff
@@ -181,19 +222,88 @@ const Stmt* SmackRep::memset(const llvm::MemSetInst& msi) {
   stringstream name;
   vector<const Expr*> args;
   name << "$memset." << region;
-  for (unsigned i = 0; i < msi.getNumOperands() - 1; i++)
+  args.push_back(expr(msi.getOperand(0)));
+  args.push_back(expr(msi.getOperand(1)));
+  const llvm::Value* setSize = msi.getOperand(2);
+  args.push_back((getIntSize(setSize) == 32)? Expr::fn("$zext.i32.ref", expr(setSize)): expr(setSize));
+  for (unsigned i = 3; i < msi.getNumOperands() - 1; i++)
     args.push_back(expr(msi.getOperand(i)));
   return Stmt::call(name.str(),args);
 }
 
-const Expr* SmackRep::pa(const Expr* base, int index, int size) {
-  return pa(base, Expr::lit(index), Expr::lit(size));
+const Stmt* SmackRep::load(const llvm::Value* addr, const llvm::Value* val) {
+// The tricky part is that we could expr(li) is actually expr(val) so that it is possible to pass li to val. 
+  const Expr* rhs = mem(addr);
+
+  if (isFloat(val))
+    rhs = Expr::fn("$si2fp", rhs);
+
+  if (!SmackOptions::BitVectors || (SmackOptions::InferFieldOverlap && isFieldDisjoint(addr, llvm::cast<const llvm::Instruction>(val))))
+    return Stmt::assign(expr(val), rhs);
+  else {
+    stringstream name;
+    name << "$load." << int_type(getElementSize(addr));
+    return Stmt::assign(expr(val), Expr::fn(name.str(), Expr::id(memPath(getRegion(addr), 8)), expr(addr)));
+  }
+}
+
+const Stmt* SmackRep::store(const llvm::Value* addr, const llvm::Value* val, const llvm::StoreInst* si) {
+// Having a default value of si (NULL) is unsound.
+  const Expr* rhs = expr(val);
+
+  if (isFloat(val))
+    rhs = Expr::fn("$fp2si", rhs);
+
+  if (!SmackOptions::BitVectors || (SmackOptions::InferFieldOverlap && isFieldDisjoint(addr, si)))
+    return Stmt::assign(mem(addr),rhs);
+  else
+    return storeAsBytes(getRegion(addr), getElementSize(addr), expr(addr), expr(val));
```

Comment:
Good catch! I almost miss it. 

`val` type and `addr`'s element type should be consistent for `load`/`store` instruction. In this case, both of them is floating point. I think all we need here is a function mapping floating point to int. The potential problem is that variables are stored to different memories depending on their size. So if the floating point size is 80bit, then there is no declared memory variables for it.